app/src/main/java/petermcneil/musiclibrary/controllers/ArtistController.java

```diff
@@ -54,19 +55,42 @@ public String getArtistList(Model model){
 
     @RequestMapping(value = "/artist", method = RequestMethod.POST)
     public String postArtist(MutableArtist muteArtist){
-
-        Artist artist = Artist.artistBuilder()
-                .name(muteArtist.getName())
-                .type(muteArtist.getType())
-                .bio(Bio.bioBuilder().biography(muteArtist.getBio()).build())
-                .build();
-
-        LOG.info("REQUEST : POST the artist {} to the library", artist.getName());
-        Integer artistId = db.post(artist);
-
+        Integer artistId = muteArtist.getArtistId();
+
+        if(artistId == null) {
+            Bio bio = Bio.bioBuilder()
+                    .biography(muteArtist.getBio())
+                    .build();
+
+            Artist artist = Artist.artistBuilder()
+                    .name(muteArtist.getName())
+                    .type(muteArtist.getType())
+                    .bio(bio)
+                    .build();
+
+            LOG.info("REQUEST : POST the artist {} to the library", artist.getName());
+            artistId = db.post(artist);
+        }else {
+            Bio bio = Bio.bioBuilder()
+                    .bioId(muteArtist.getBioId())
+                    .biography(muteArtist.getBio())
+                    .build();
+
+
+            Artist artist = Artist.artistBuilder()
+                    .artistId(artistId)
+                    .name(muteArtist.getName())
+                    .type(muteArtist.getType())
+                    .bio(bio)
+                    .build();
+
+            LOG.info("REQUEST : PUT the artist ({}), with id ({}) into the library", artist.getName(), artistId);
+            db.put(artist, artistId);
+        }
         return "redirect:/artist/" + artistId;
     }
 
+    //TODO Get rid of this?
```

Comment:
usually for CRUD services, when POSTing you are adding a new entity, and when PUTing you are updating an already existing entity (given the id of the entity), it seems you are trying to do both in the POST here. I'm guessing this might be because of tymeleaf?